### PR TITLE
Making SqlServer work with triggers

### PR DESCRIPTION
**Pull request by lorenzo**
This is a fix for #7765.

### CODE REVIEW

**Comment by lorenzo** [comment]
@stmeyer I forced pushed your branch, The easiest way to get back the code is to re-clone your fork so you get my changes

File: src/Database/Dialect/SqlserverDialectTrait.php [hunk 1]
```diff
@@ -325,6 +325,10 @@ public function rollbackSavePointSQL($name)
      */
     public function newCompiler()
     {
+        if (isset($this->_config['compiler'])) {
```
Comment (markstory) [inline src/Database/Dialect/SqlserverDialectTrait.php]: This is a fairly low level option. Should we provide any abstractions here? For example

```
'compiler' => 'default | triggers'
```

File: src/Database/SqlserverWithTriggerCompiler.php [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        $driver = $db->driver();
+        $dialect = $driver->schemaDialect();
+        $collection = $db->schemaCollection();
+        $description = $collection->describe($tableWithoutSchema);
+        $constraints = $description->constraints();
+        $primaryKeyColumns = [];
+        foreach ($constraints as $constraintName) {
+            $constraint = $description->constraint($constraintName);
+            if ($constraint['type'] === \Cake\Database\Schema\Table::CONSTRAINT_PRIMARY) {
+                $primaryKeyColumns = $constraint['columns'];
```
Comment (markstory) [inline src/Database/SqlserverWithTriggerCompiler.php]: You can get the primary key columns with `$description->primaryKey()`

File: src/Database/SqlserverWithTriggerCompiler.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        }
+        $sqlCreateColumns = [];
+        $primaryKeyColumnsForSql = [];
+        foreach ($description->columns() as $columnName) {
+            if (in_array($columnName, $primaryKeyColumns)) {
+                $descriptionOfSingleColumn = $description->column($columnName);
+                if (isset($descriptionOfSingleColumn['autoIncrement']) && $descriptionOfSingleColumn['autoIncrement']) {
+                    $descriptionOfSingleColumn['autoIncrement'] = false;
+                }
+                $description->addColumn($columnName, $descriptionOfSingleColumn);
```
Comment (markstory) [inline src/Database/SqlserverWithTriggerCompiler.php]: Do these mutations to the schema object need to be rolled back ever?

File: appveyor.yml [hunk 4]
```diff
@@ -85,3 +85,5 @@ test_script:
   - sqlcmd -S ".\SQL2012SP1" -U sa -P Password12! -Q "create database cakephp;"
   - cd C:\projects\cakephp
   - vendor\bin\phpunit.bat
+  - $env:db_dsn = "sqlserver://sa:Password12!@.\SQL2012SP1/cakephp?MultipleActiveResultSets=false&compiler=Cake\DatabaseSqlserverWithTrigger"
+  - vendor\bin\phpunit.bat --testsuite "Database Test Suite"
```
Comment (markstory) [inline appveyor.yml]: Clever way to not need to write more tests :smile:

Comment (antograssiot) [inline appveyor.yml]: @lorenzo have you tried to add this as a second matrix configuration ?

Comment (lorenzo) [inline appveyor.yml]: No I have not... would you like to help me with that @antograssiot ?

File: src/Database/SqlserverWithTriggerCompiler.php (hunk 2 above)
Comment (stmeyer) [inline src/Database/SqlserverWithTriggerCompiler.php]: Shure, but I need to remove the constraint from the description, see next line (line 57).

**Comment by stmeyer** [comment]
@lorenzo @antograssiot any news about this? I am waiting to start this in production :smile:

**Comment by antograssiot** [comment]
@stmeyer I think that the appveyor build can be added by removing the modification made to the appveyor.yml and replace the matrix part in the top of the file by this:
```yml

  matrix:
      - db: 2012
        db_dsn: 'sqlserver://sa:Password12!@.\SQL2012SP1/cakephp?MultipleActiveResultSets=false'

      - db: 2012
db_dsn: 'sqlserver://sa:Password12!@.\SQL2012SP1/cakephp?MultipleActiveResultSets=false&compiler=Cake\Database\SqlserverWithTrigger'
```

it should trigger a second complete build to see if all tests passes and can be improved later to reduce the build time.

**Comment by markstory** [comment]
@stmeyer We need to get a passing build with sqlserver before this can be merged in.

**Comment by stmeyer** [comment]
@antograssiot thanks for the configuration, it's running a second test now, but breaks with tables without primary key. I will fix it.

**Comment by stmeyer** [comment]
I have tested it, and it is not working, and it shoud never have worked. The PR is generating SQL Server Input statement like the following:
```php
$connection = ConnectionManager::get('default');
$results = $connection->execute("
DECLARE @var TABLE ([id] INTEGER NOT NULL);

INSERT INTO example (
	col1
	,col2
	,created
	,modified
	,created_by_user_id
	,modified_by_user_id
	)
OUTPUT INSERTED.id
INTO @var
VALUES (
	'exampleText1'
	,'exampleText2'
	,'2016-03-02 13:34:32'
	,'2016-03-02 13:34:32'
	,1
	,1
	);

SELECT *
FROM @var
");
debug($results->fetchAll(\PDO::FETCH_ASSOC));
```
The example throws the error: `Error: [PDOException] SQLSTATE[IMSSP]: The active result for the query contains no fields.`.

The same statement is working with `OUTPUT`:
```php
$connection = ConnectionManager::get('default');
$results = $connection->execute("
INSERT INTO example (
	col1
	,col2
	,created
	,modified
	,created_by_user_id
	,modified_by_user_id
	)
OUTPUT INSERTED.*
VALUES (
	'exampleText1'
	,'exampleText2'
	,'2016-03-02 13:34:32'
	,'2016-03-02 13:34:32'
	,1
	,1
	);
");
debug($results->fetchAll(\PDO::FETCH_ASSOC));

// Result:
[
	(int) 0 => [
		(int) 'id' => '439',
		(int) 'col1' => '6',
		(int) 'col2' => 'error_codes',
		(int) 'created' => '2016-03-02 13:34:32.000',
		(int) 'modified' => '2016-03-02 13:34:32.000',
		(int) 'created_by_user_id' => '1',
		(int) 'modified_by_user_id' => '1'
	]
]
```

The `OUTPUT.*` was used, due to fix https://github.com/cakephp/cakephp/issues/5104. Maybe it would be an other fix to work with triggers, by making the `OUTPUT.*` configuratable, so that Cake implements 2 different SQL Server compilers. The default one compiles with `OUTPUT.*` but if someone is not using lower transaction isolation levels, he could configure the second compiler.

How do you think about?

**Comment by lorenzo** [comment]
@stmeyer that would not work, we need a safe way of returning the last inserted id. Another idea is to use your code conditionally and the `OUTPUT *` conditionally as well, depending or not on the presence of a primary key

**Comment by stmeyer** [comment]
@lorenzo sorry, i do not know how exactly you want to use `OUTPUT *`. If you can explain it, i will try to change this PR:

**Comment by stmeyer** [comment]
Any news here? It is still not possible to use CakePHP 3.X with Microsoft sql server if replication is enabled. A new driver for sql server has been released for PHP 7: https://github.com/Azure/msphpsql/releases. Has someone checked, if the new driver fixed #5104?

**Comment by lorenzo** [comment]
No news here on my part. Were you using your own patch?

**Comment by stmeyer** [comment]
I'm not using this PR, too. As written in a other comment, it could not run. I'm using the following workaround. This removes the `OUTPUT` from the `INSERT` query. I am running with PHP7 and SqlSrv 4 RTW. I have no problems with saving entities.
```php
Config:
'Datasources' => [
        'default' => [
            'className' => 'Cake\Database\Connection',
            'driver' => 'App\Database\Driver\Sqlserver',
            'compiler' => 'App\Database\SqlserverWithTrigger',

/**********************************************/
App\Database\Driver\Sqlserver:

class Sqlserver extends \Cake\Database\Driver\Sqlserver
{
    /**
     * {@inheritDoc}
     *
     * @return \Cake\Database\SqlserverCompiler
     */
    public function newCompiler()
    {
        if (isset($this->_config['compiler'])) {
            $class = $this->_config['compiler'] . 'Compiler';
            return new $class();
        }
        return new SqlserverCompiler();
    }
}
/**********************************************/
App\Database\SqlserverWithTriggerCompiler:
class SqlserverWithTriggerCompiler extends SqlserverCompiler
{
    /**
     * Generates the INSERT part of a SQL query
     *
     * SQL Server with enabled triggers does not allow the OUTPUT clause
     * so that the INSERT is generated with a temporary variable.
     *
     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
     *
     * @param array $parts The parts to build
     * @param \Cake\Database\Query $query The query that is being compiled
     * @param \Cake\Database\ValueBinder $generator the placeholder generator to be used in expressions
     * @return string
     */
    protected function _buildInsertPart($parts, $query, $generator)
    {
        $table = $parts[0];
        $columns = $this->_stringifyExpressions($parts[1], $generator);
        return sprintf('INSERT INTO %s (%s) ', $table, implode(', ', $columns));
    }
}
```

**Comment by lorenzo** [comment]
Ok, I'll close this pull request then. Any chance you can publish that code as a plugin?

**Comment by xartviel** [comment]
I am having a problem with composite primary key. Is there anyone who also encountered this issue?

**Pull request closed without merging**